File: 0916-word-subsets/0916-word-subsets.py
class Solution:
    def wordSubsets(self, words1: List[str], words2: List[str]) -> List[str]:
        freq = [0 for i in range(26)]
        result = list()

        for word in words2:
            _freq = [0 for i in range(26)]
            for letter in word:
                _freq[ord(letter) - ord("a")] += 1
            for i in range(26):
                if freq[i] < _freq[i]:
                    freq[i] = _freq[i]

        for word in words1:
            _freq = [0 for i in range(26)]
            for letter in word:
                _freq[ord(letter) - ord("a")] += 1
            for i in range(26):
                if _freq[i] < freq[i]:
                    break
            else:
                result.append(word)

        return result

File: 0916-word-subsets/0916-word-subsets.java
class Solution {
    public List<String> wordSubsets(String[] words1, String[] words2) {
        int[] freq = new int[26];
        List<String> result = new ArrayList<>();

        for (String word: words2) {
            int[] f = new int[26];
            for (int i = 0; i < word.length(); i++) {
                f[word.charAt(i) - 'a']++;
            }
            for (int i = 0; i < 26; i++) {
                if (freq[i] < f[i]) {
                    freq[i] = f[i];
                }
            }
        }

        for (String word: words1) {
            int[] f = new int[26];
            boolean flag = true;
            for (int i = 0; i < word.length(); i++) {
                f[word.charAt(i) - 'a']++;
            }
            for (int i = 0; i < 26; i++) {
                if (f[i] < freq[i]) {
                    flag = false;
                    break;
                }
            }
            if (flag) result.add(word);
        }

        return result;
    }
}